New Maxpool2D Op

mywoodstock opened the pull request:
### Note
Reviewers please note that this is PR#1 of 2, here I have separated out the new implementation completely from the old version into separate source files. The new stuff are in `max_pool2d*` sources. In the next PR, I will remove the old version completely (i.e. `rm max_pool_*`.)

### Ticket
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/9999)

### Problem description
New implementation of jit maxpool2d and using new TTNN infra.

### Checklist
- [x] Post commit CI passes
- [x] Model regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

mywoodstock committed:
#9999: New jit maxpool2d, with latest ttnn structure. old maxpool still
exists and works, but will completely remove it in next PR.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/maxpool_pybind.hpp:
```diff
@@ -9,13 +9,14 @@
 
 #include "ttnn/cpp/pybind11/decorators.hpp"
```
Please split to .hpp and .cpp
This is the best example we got so far
https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/ccl/all_gather/all_gather_pybind.hpp

This will help keep build time lower

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d_pybind.hpp:
```diff
@@ -0,0 +1,81 @@
+
+#pragma once
+
+#include <pybind11/pybind11.h>
```
https://github.com/tenstorrent/tt-metal/pull/10825/files#r1697623357

ayerofieiev-tt replied:
can you please check ccl/all_gather binding header? it shows how to avoid these heavy includes

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.hpp:
```diff
@@ -0,0 +1,29 @@
+
+struct MaxPoolNewOp {
+    template<typename T>
+    static Tensor operator()(uint8_t queue_id, const Tensor& input_tensor, uint32_t batch_size, uint32_t input_h, uint32_t input_w, uint32_t channels, std::array<uint32_t, 2> kernel_size, std::array<uint32_t, 2> stride, std::array<uint32_t, 2> padding, std::array<uint32_t, 2> dilation, T* device);
```
I would love to see these args compressed into a struct. 
You will have to add Struct bindings for it to work too. That will keep options list sane
Wdyt?

mywoodstock replied:
I would keep them as is because:
1. it matches the pytorch maxpool args list
2. we will eventually get rid of `batch_size`, `input_h`, `input_w` and `channels` args once the new tensor infra is ready.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.hpp:
```diff
@@ -0,0 +1,29 @@
+
+struct MaxPoolNewOp {
+    template<typename T>
+    static Tensor operator()(uint8_t queue_id, const Tensor& input_tensor, uint32_t batch_size, uint32_t input_h, uint32_t input_w, uint32_t channels, std::array<uint32_t, 2> kernel_size, std::array<uint32_t, 2> stride, std::array<uint32_t, 2> padding, std::array<uint32_t, 2> dilation, T* device);
```
This is not typical for a ttnn operation today, but we discussed with the team recently that we expect ops to potentially gain this parameter overtime.

@mywoodstock  can you share more about this argument `T* device`? 

CC @eyonland @arakhmati @dmakoviichuk-tt

mywoodstock replied:
Yes, so maxpool (and conv as well), use the halo construction infra which need the device to obtain the physical noc coords to be used in the op -- and it is template'ized to facilitate multi-device use.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp:
```diff
@@ -0,0 +1,109 @@
+        MaxPoolNew::tensor_args_t{.input_tensor_ = haloed_tensor});
+}
+
+// device template specializations
+template Tensor MaxPoolNewOp::operator()<Device>(uint8_t queue_id, const Tensor& input_tensor, uint32_t batch_size, uint32_t input_h, uint32_t input_w, uint32_t channels, std::array<uint32_t, 2> kernel_size, std::array<uint32_t, 2> stride, std::array<uint32_t, 2> padding, std::array<uint32_t, 2> dilation, Device* device);
+template Tensor MaxPoolNewOp::operator()<DeviceMesh>(uint8_t queue_id, const Tensor& input_tensor, uint32_t batch_size, uint32_t input_h, uint32_t input_w, uint32_t channels, std::array<uint32_t, 2> kernel_size, std::array<uint32_t, 2> stride, std::array<uint32_t, 2> padding, std::array<uint32_t, 2> dilation, DeviceMesh* device);
```
❤️

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool2d.cpp:
```diff
@@ -0,0 +1,109 @@
+        .memory_config_ = memory_config};
+
+    // and then call the maxpool uop
+    return ttnn::device_operation::run<MaxPoolNew>(
```
Following the new TMP infra! Nice! 👍

tt-aho committed:
#9999: Fix multi-device for max_pool2d
Add pybinding for DeviceMesh overload for max_pool2d
Switch to register_operation_with_auto_launch_op for handling DeviceMesh

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.hpp:
```diff
@@ -0,0 +1,83 @@
+    static tt::stl::hash::hash_t compute_program_hash(const operation_attributes_t&, const tensor_args_t&);
+
+    // call old funcs from the above
+    static void validate(const Tensor& input, const tt::tt_metal::SlidingWindowConfig& sliding_window_config, const MemoryConfig& out_mem_config);
+    static shape_return_value_t compute_output_shapes(const Tensor& input, const tt::tt_metal::SlidingWindowConfig& sliding_window_config, DataType output_dtype, const MemoryConfig& out_mem_config);
+    static tensor_return_value_t create_output_tensors(const Tensor &input, const tt::tt_metal::SlidingWindowConfig& sliding_window_config, DataType output_dtype, const MemoryConfig& out_mem_config);
+    static operation::OpPerformanceModel create_op_performance_model(const operation_attributes_t&, const tensor_args_t&, const Tensor&);
```
these can go to .cpp, right? no need to expose them in the header

mywoodstock replied:
yes, let me just get rid of these -- no point in trying to keep old function signatures.

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.hpp:
```diff
@@ -0,0 +1,83 @@
+
+        using cached_program_t = ttnn::device_operation::CachedProgram<shared_variables_t>;
+
+        static cached_program_t create(const operation_attributes_t& operation_attributes,
+                                       const tensor_args_t& tensor_args,
+                                       tensor_return_value_t& output_tensor);
+        static void override_runtime_arguments(cached_program_t& cached_program,
+                                               const operation_attributes_t& operation_attributes,
+                                               const tensor_args_t& tensor_args,
+                                               tensor_return_value_t& output_tensor);
+        static cached_program_t max_pool_2d_multi_core_sharded_with_halo_v2_new(const Tensor &input,
```
These are defined in device_operation header, but not implemented in device_operation.cpp
My preference is to keep declaration and definition aligned between files unless absolutely necessary to do otherwise.

@arakhmati , as far as I understand this is based on the example?

ayerofieiev-tt replied:
I expect this to be declared in `MultiCoreMaxPoolProgramFactory.hpp`, implemented in `MultiCoreMaxPoolProgramFactory.cpp` and if we want to avoid including that header here - it can be forward declared.

My thinking here is next:
There is no need for consumers of MaxPool2D Device Operation to be aware about these programs. 
They pretty much a detail of implementation of the device operation itself.

mywoodstock replied:
I got rid of this function, so no longer here :)

mywoodstock committed:
#9999: Move pybind impl to cpp

tt-aho commented on tests/ttnn/unit_tests/operations/test_maxpool2d.py:
```diff
@@ -186,7 +193,11 @@ def test_run_max_pool(
+    # torch.save(golden_pytorch, "golden_pytorch.pt")
+    passing, pcc = assert_with_pcc(output_pytorch, golden_pytorch)
+
+    print(f"Passing: {passing}, PCC: {pcc}")
```
Use loguru instead of print

tt-aho commented on ttnn/ttnn/operations/conv2d.py:
```diff
@@ -392,10 +394,14 @@ def get_grid_size_and_num_cores_nhw_from_core_grid(core_grid, height_sharded):
 
 # internal helper function. not exposed to user.
 def create_sharded_memory_config_from_parallel_config(tensor_shape, parallel_config, tile_size):
+    logger.info(
+        f"py create_sharded_memory_config_from_parallel_config: {tensor_shape}, {parallel_config.num_cores_nhw} {parallel_config.grid_size}, {tile_size}"
+    )
```
Make this a debug instead of info print

ayerofieiev-tt commented on ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sliding_window_op_infra/halo_op.hpp:
```diff
@@ -34,7 +35,7 @@ struct Halo {
 
     static constexpr auto attribute_names =
-        std::make_tuple("config_", "parallel_config_", "pad_val_", "remote_read_", "transpose_mcast_", "reshard_num_cores_nhw_", "max_out_nsticks_per_core_", "output_memory_config_");
+        std::make_tuple("config_", "parallel_config_", "pad_val_", "remote_read_", "transpose_mcast_", "reshard_num_cores_nhw_", "max_out_nsticks_per_core_", "output_memory_config_", "is_out_tiled_");
```
can you please check with @arakhmati , I think this can be removed now

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool.hpp:
```diff
@@ -12,13 +12,13 @@
+namespace tt {
+namespace tt_metal {
+
 inline uint32_t ceil_multiple_of(uint32_t n, uint32_t m) {
```
same function is declared in ttnn/cpp/ttnn/operations/pool/maxpool/device/max_pool2d_device_op.hpp

mywoodstock replied:
Ah yes, this file will be deleted next :)

ayerofieiev-tt commented on ttnn/cpp/ttnn/operations/pool/maxpool/max_pool.hpp:
```diff
@@ -12,13 +12,13 @@
     return (uint32_t) std::ceil((float) n / m) * m;
 }
 
-namespace tt {
```
this file will be removed next?

TT-BrianLiu reviewed (approved):
Minor comment

pavlejosipovic commented on ttnn/cpp/ttnn/deprecated/tt_dnn/op_library/sliding_window_op_infra/sliding_window.cpp:
```diff
@@ -359,19 +364,28 @@ namespace tt::tt_metal::sliding_window {
 
-    std::vector<uint16_t> flatten(const std::vector<std::vector<uint16_t>>& input) {
+    std::vector<uint16_t> flatten(const std::vector<std::vector<uint16_t>>& input, uint32_t extend_with_zeroes) {
         std::vector<uint16_t> flattened_vector;
```
maybe reserve enough space here to avoid mulitple allocations?
